src/ZODB/mvccadapter.py

```diff
@@ -0,0 +1,268 @@
+        'getName', 'getSize', 'history', 'lastTransaction', 'sortKey',
+        'loadBlob', 'openCommittedBlobFile',
+        )
+
+    def __init__(self, storage):
+        self._storage = storage
+        for name in self._copy_methods:
+            if hasattr(self._storage, name):
+                setattr(self, name, getattr(storage, name))
+
+        if interfaces.IBlobStorage.providedBy(storage):
+            zope.interface.alsoProvides(self, interfaces.IBlobStorage)
+
+    def isReadOnly(self):
+        return self._storage.isReadOnly()
+
+    def supportsUndo(self):
+        try:
+            return self._storage.supportsUndo()
+        except AttributeError:
+            return False
+
+    def undoLog(self, first, last, filter=None):
+        try:
+            return self._storage.undoLog(first, last, filter)
+        except AttributeError:
+            return False
+
+    def undoInfo(self, first=0, last=20, specification=None):
+        try:
+            return self._storage.undoInfo(first, last, specification)
+        except AttributeError:
+            return False
+
+    def temporaryDirectory(self):
+        return self._storage.temporaryDirectory()
+
+    def __len__():
+        return len(self._storage)
+
+class MVCCAdapter(Base):
+
+    def __init__(self, storage):
+        Base.__init__(self, storage)
+        self._instances = set()
+        self._lock = threading.Lock()
+        if hasattr(storage, 'registerDB'):
+            storage.registerDB(self)
+
+    def new_instance(self):
+        instance = MVCCAdapterInstance(self)
+        with self._lock:
+            self._instances.add(instance)
+        return instance
+
+    def before_instance(self, before=None):
+        return BeforeAdapterInstance(self, before)
+
+    def undo_instance(self):
+        return UndoAdapterInstance(self)
+
+    def _release(self, instance):
+        with self._lock:
+            self._instances.remove(instance)
+
+    closed = False
+    def close(self):
+        if not self.closed:
+            self.closed = True
+            self._storage.close()
+            del self._instances
+            del self._storage
+
+    def invalidateCache(self):
+        with self._lock:
+            for instance in self._instances:
+                instance._invalidateCache()
+
+    def invalidate(self, transaction_id, oids, version=''):
+        with self._lock:
+            for instance in self._instances:
+                instance._invalidate(oids)
+
+    def _invalidate_finish(self, oids, committing_instance):
+        with self._lock:
+            for instance in self._instances:
+                if instance is not committing_instance:
+                    instance._invalidate(oids)
+
+    references = serialize.referencesf
+    transform_record_data = untransform_record_data = lambda self, data: data
+
+    def pack(self, pack_time, referencesf):
+        return self._storage.pack(pack_time, referencesf)
+
+class MVCCAdapterInstance(Base):
```

Comment:
> Do you know of code in the [wild?] that would need this?

There are maybe 20 RelStorage tests that rely on something like `conn._storage.needs_poll()` and so forth. Now, those won't directly be affected because relstorage already implements IMVCCStorage, but it seemed like if RS did it, possibly others did too. (That does highlight a difference though, in the handling of IMVCCStorages vs not: you lose access to custom attributes in the "not" case that you used to be able to get.) I can imagine test code directly trying to get to the dicts maintained by a MappingStorage, for instance.

In NextThought's code, I found at least one place where we access `conn._storage.load`. That may be a bug, and it wouldn't be affected by this because the adapter provides `load` (but not `loadBefore` AFAICS), but here too, it seemed like if we did it, probably others did too.

Of course, the plural of "anecdote" is not data, and personally I tend to be conservative about potentially breaking changes even in implementation details.